### ISSUE TITLE
Question about 32000 section 7.9.4: Dates with no timezone offset information

### ISSUE THREAD

**Issue by datalogics-pgallot** [pdf-association/pdf-issues#251]
If I have:
D:20221128110300
is this functionally equivalent to 
D:20221128110300Z
or is it dependent on my local time zone?

Section 7.9.4 has the following text:
"the LATIN CAPITAL LETTER Z signifies that local time is equal to UT. If no UT information is specified, the relationship of the specified time to UT shall be considered to be GMT. Regardless of whether the time zone is specified, the rest of the date shall be specified in local time."

I ask because I'm looking at a bug report where the current behavior leans on the latter part of this statement to treat "D:20221128110300" as 2022.11.28 11:03:00 -06'00' on my machine, while I interpret the (somewhat muddled) middle part of the statement as indicating it shall be treated as: 2022.11.28 11:03:00 +00'00'

Specifically, the intent of this sentence:
"If no UT information is specified, the relationship of the specified time to UT shall be considered to be GMT" 
seems to be: 
"If no UT information is specified, the missing timezone offset shall be assumed to be the same as Greenwich Mean Time's  timezone offset (+0'00)."
But that isn't quite what it says.

**Comment by lrosenthol** [comment]
This has come up in the context of PDF/A & XMP a number of times over the years.  Recommend that if we do anything in 32K, it needs to be based on the resolutions in PDF/A.

**Comment by petervwyatt** [comment]
16684-1:2019 Dates state:
> The time zone designator need not be present in XMP. When not present, the time zone is unknown, and an XMP processor should not assume anything about the missing time zone.
> Local time-zone designators +hh:mm or –hh :mm should be used when possible instead of converting to UTC.
NOTE If a file was saved at noon on October 23, a timestamp of 2004-10-23T12:00:00-06:00 conveys more information than 2004-10-23T18:00:00Z.

**Comment by DietrichSeggern** [comment]
The full para in Section 7.9.4 reads:
`The prefix “D:” shall be present, the year field (YYYY) shall be present and all other fields may be present but only if all of their preceding fields are also present. The APOSTROPHE following the hour offset field (HH) shall only be present if the HH field is present. The minute offset field (mm) shall only be present if the APOSTROPHE following the hour offset field (HH) is present. The default values for MM and DD shall be both 01; all other numerical fields shall default to zero values. A PLUS SIGN as the value of the O field signifies that local time is now and later than UT, a HYPHEN-MINUS signifies that local time is earlier than UT, and the LATIN CAPITAL LETTER Z signifies that local time is equal to UT. If no UT information is specified, the relationship of the specified time to UT shall be considered to be GMT. Regardless of whether the time zone is specified, the rest of the date shall be specified in local time.`

"all other numerical fields shall default to zero values" includes the time zone offset which would mean that GMT is the default if there is no entry.

For PDF/A we came to the conclusion that the same defaults that are specified above for date strings in PDF need to be applied to XMP. There is no provision that would make that explicit but in an example in PDF/A-1 (section 6.7.3) it is made clear that these entries are equivalent:
`/CreationDate (D:20040402)`
`<xmp:CreateDate>2004-04-02</xmp:CreateDate>`
That is only possible when the default rules are the same in both cases.

**Comment by petervwyatt** [comment]
So based on the quoted text from XMP (ISO 16684-1:2019) and the quoted text from PDF it seems that they have specified different default rules! XMP defaults ("should") to no assumptions while PDF defaults to GMT via a "shall" statement.

So is PDF/A requiring additional wording over the XMP stated "should"?

**Comment by DietrichSeggern** [comment]
I think that we for PDF/A agreed to the following: 
- This document (that is also referenced in the Bibliography of ISO 16684-1:2019) specifies that a standard may define a "granularity" specifying the "the meaning of the dates and times with reduced precision" (in the section "Formats").
- PDF/A does not explicitly specify such "granularity", however it has examples that make clear that the intention is that date defaults for DocInfo and XMP are the same:
`/CreationDate (D:20040402)
<xmp:CreateDate>2004-04-02</xmp:CreateDate>`
Therefore we should use the same defaults in both cases.

**Comment by faceless2** [comment]
I agree with Leonard that the approach from XMP - where no timezone means "unspecified, make no assumption" rather than "presume UTC" - is definitely preferred.

I agree with Dietrich that specifically for PDF/A this is not really significant anymore - it matters for PDF/A-1, but that's for the PDF/A-1 WG to resolve(*). It doesn't matter for any other revision of PDF/A as the Info dictionary is either unused or (for PDF/A-4) disallowed due to deprecation.

However the PDF Date syntax is still used for Annotation and EmbeddedFile creation dates, so Patricks's question still needs resolving.

Given it still matters and the topic is open, I would like to partially hijack this issue to suggest that as part of any resolution here, we do something about the "trailing apostrophe problem", which with uncanny timing I wrote an article on at https://bfo.com/blog/2023/01/17/odds_and_ends_dates_timezones_and_apostrophes/ recently.

Briefly:

* PDF 1.6 and earlier format dates like`D:20210313144715+01'00'`
* PDF 1.7 and ISO32000 format dates like`D:20210313144715+01'00`

The difference is the trailing apostrophe.

Unfortunately the older syntax is both a) generated and b) required for correct parsing in "a large PDF product" - as such it remains with us many years after it was supposed to have changed. To demonstrate, [here is a PDF](https://github.com/pdf-association/pdf-issues/files/10958260/with-apostrophe.pdf)
which has creation and modified dates at midnight UTC, but in the Indian timezone of +05:30 - the creation date has a trailing apostrophe (old-syntax), and the modification date does not (new syntax).

To reflect reality I'd propose that the trailing apostrophe on dates is noted as optional: anyone parsing these has to handle both syntaxes, and anyone generating them has to use the older syntax or risk questions from customers in India, Adelaide or other timezones with a non-zero minute offset.


(*) it's complicated because the revision of XMP current at the time of PDF/A-1 _required_ timezones for any dates with a time component, a requirement not in later revisions.

**Comment by datalogics-pgallot** [comment]
The use of CreationDate and ModDate in the Info dict is *not* deprecated in PDF 2.0 (see section 14.3.3), so having them the same in the XMP metadata should possibly still be a concern.

I did a rough survey of where 7.9.4 dates are used in the 32000-2:2020: Info Dict, Embedded Files, Collections, Threads(?), Annotations, Digital Signatures and VRI dictionaries, PagePiece Dictionaries, Web Capture, and TrapNetwork Annotations.

I don't know if 'no timezone means "unspecified, make no assumption"' is entirely appropriate. If there's no timezone, then the timestamp date is only accurate to +/- 1 day of the date-portion of the timestamp. but if you are comparing timestamps within a homogeneous set of timestamps, then more precise differences between the time portions of the time stamps might be meaningful.

**Comment by DietrichSeggern** [comment]
PDF/A-1 and PDF/A-2 have the explicit requirement that corresponding metadata entries in Document Information and in XMP need to have equivalent values. This requirement is not present in PDF/A-4.

**Comment by faceless2** [comment]
It's a bit of a subtle distinction - the Info dictionary as a whole is deprecated (in section 0.3, in the introduction for ISO320002:2020), and according to table 15 "this method for specifying document metadata has been deprecated in PDF 2.0 and should therefore only be used to encode information that is stated as required elsewhere in this document."

As it happens that includes the `ModDate` if (and so far, only if) Page-Piece dictionaries are required, so I do sort of concede your point. I definitely agree Dates are widely used elsewhere, the Arlington Model has them present in 68 tables.

I think my point is that the only place the specification requires us to compare PDF dates (where no timezone means UTC) and XMP dates (where no timezone means floating) is PDF/A-1 - and the XMP spec from that era disallowed dates without timezones. In all later versions of PDF/A "_a compliant reader SHALL ignore [the Info Dictionary]_"). And for any other dates used in PDF, if they are only accurate to with 1 day then that's annoying, sure, but without a specification requirement to compare them to dates specified elsewhere, I'm not sure it's a spec issue.

**Comment by faceless2** [comment]
> "A document information dictionary may appear within a conforming file. If it does appear, **a compliant PDF/A-2 reader shall ignore it**. A PDF/A-2 conforming writer should ensure that the values in the document information dictionary are consistent with the corresponding values in the document’s metadata stream as listed in Table 7."

Readers must ignore it, writers must ensure they're consistent.

**Comment by DietrichSeggern** [comment]
Thanks, you are obviously right.

> Readers must ignore it, writers SHOULD ensure they're consistent.

**Comment by petervwyatt** [comment]
Comment: PDF/A dated revision has an opportunity to clarify.

@datalogics-pgallot - was your context PDF/A or general PDF?

**Comment by datalogics-pgallot** [comment]
@petervwyatt It's been a while, but I'm pretty sure that my context for bringing this up was general PDF

**Comment by petervwyatt** [comment]
Thanks. 

So the focus of this errata is to discuss what might/can/should be stated in only ISO 32000-2 (PDF 2.0) about date objects with no timezone.

**Comment by petervwyatt** [comment]
Trying to circle back around and finally conclude this errata for PDF 2.0 (ISO 32000-2:2020), since PDF/A is fully addressed (_as per discussions above_):

@DietrichSeggern's [Mar 10 comment](https://github.com/pdf-association/pdf-issues/issues/251#issuecomment-1463660858) highlights "all other numerical fields shall default to zero values" includes the time zone offset which would mean that GMT is the default if there is no entry.

so therefore @datalogics-pgallot's original suggestion seems an appropriate improvement to me:

Current wording:
> If no UT information is specified, the relationship of the specified time to UT shall be considered to be GMT.

Proposed solution wording:
> If no UT information is specified, the missing timezone offset shall be assumed to be the same as Greenwich Mean Time's timezone offset (+0'00).

**Comment by petervwyatt** [comment]
PDF TWG agree
